File: security_assistant/remediation/code_examples/path_traversal_fix.go
// VULNERABLE:
// filepath := "/var/www/uploads/" + filename
// data, err := ioutil.ReadFile(filepath)

// SECURE (Path Validation with filepath.Clean):
import (
    "path/filepath"
    "strings"
)

baseDir := "/var/www/uploads"
cleanPath := filepath.Clean(filepath.Join(baseDir, filename))

// Ensure path is within base directory
if !strings.HasPrefix(cleanPath, baseDir) {
    return errors.New("path traversal attempt detected")
}

data, err := ioutil.ReadFile(cleanPath)

File: security_assistant/remediation/code_examples/path_traversal_fix.rs
// VULNERABLE:
// let path = format!("/var/www/uploads/{}", filename);
// let contents = fs::read_to_string(path)?;

// SECURE (Path Canonicalization and Validation):
use std::path::{Path, PathBuf};
use std::fs;

let base_dir = Path::new("/var/www/uploads").canonicalize()?;
let requested_path = base_dir.join(filename).canonicalize()?;

// Ensure path is within base directory
if !requested_path.starts_with(&base_dir) {
    return Err("Path traversal attempt detected".into());
}

let contents = fs::read_to_string(requested_path)?;
